### ISSUE TITLE
How to get the unit from a state description without a space between pattern and unit?

### ISSUE THREAD

**Issue by jimtng** [openhab/openhab-core#2967]
As far as I know, a state description such as this is valid: `%d°C` i.e. without a space between `%d` and its unit `°C`

However, when trying to use UnitUtils.parseUnit to extract the unit, it failed and returned `null`. Is there another way of extracting the unit, or should UnitUtils.parseUnit be modified to handle this case?

I am trying to get the unit from an item, such as
```
Number:Temperature MyItem "Temperature [%d°C]"
```

My code:
```
UnitUtils.parseUnit(MyItem.getStateDescription().getPattern())
```

Knowing that MyItem.getState() can return an UnDefType, I cannot rely on the state to obtain the unit. This is why I am resorting to MyItem.getStateDescription().getPattern()

Finally, my ultimate goal is to return a QuantityType from Persistence data when the underlying item is a dimensioned one such as the above.

**Comment by J-N-K** [comment]
I doubt that is a good approach. Since this is only useful for `NumberItem`s, why don't you use `NumberItem.getUnit()` or NumberItem.getUnitSymbol()`;

**Comment by jimtng** [comment]
When NumberItem is NULL or UNDEF, NumberItem.getUnit will return null, whereas I would still like to apply the unit to Persistence's historic data.

**Comment by J-N-K** [comment]
The code in `NumberItem` to determine the unit is 

```
         StateDescription stateDescription = getStateDescription();
         if (stateDescription != null) {
            String pattern = stateDescription.getPattern();
            if (pattern != null) {
                Unit<?> stateDescriptionUnit = UnitUtils.parseUnit(pattern);
                if (stateDescriptionUnit != null) {
                    return stateDescriptionUnit;
                }
            }
        }
```

which is similar to your code (except the null-checks. If this code can't parse the state description. The documentation on openhab.org says "The framework assumes that the unit to parse is always the last token in the state description." so it is clear that it is expected that value and unit are separated by a space.

**Comment by jimtng** [comment]
Yes I understand what the core _currently_ needs a space in order to recognise when the format specifier stops and the unit starts. However, IMO it should be perfectly reasonable to have no space between the value and the unit, e.g. `11°C` or `15%` should it be desired.

**Comment by J-N-K** [comment]
Looking at [this official document, chapter 5.3.3](/https://www.bipm.org/documents/20126/41483022/si_brochure_8.pdf) it seems the space is required by convention:

> The numerical value always precedes the unit, and a space is always used to separate the unit from the number. Thus the value of the quantity is the product of the number and the unit, the space being regarded as a multiplication sign (just as a space between units implies multiplication). The only exceptions to this rule are for the unit symbols for degree, minute, and second for plane angle, °, ′, and ′′, respectively, for which no space is left between the numerical value and the unit symbol.
This rule means that the symbol °C for the degree Celsius is preceded by a space when one expresses values of Celsius temperature t.
Even when the value of a quantity is used as an adjective, a space is left between the numerical value and the unit symbol. Only when the name of the unit is spelled out would the ordinary rules of grammar apply, so that in English a hyphen would be used to separate the number from the unit.

I did check the code of the indriya library (that we use for UoM) and it also requires a defined delimiter between number part and unit part. IMO we should not work around wrong usage of unit symbols.
